source/credentials_utils.c

```diff
@@ -383,23 +383,108 @@ static struct aws_byte_cursor s_aws_isoe_dns_suffix = AWS_BYTE_CUR_INIT_FROM_STR
 static struct aws_byte_cursor s_isof_region_prefix = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("us-isof-");
 static struct aws_byte_cursor s_aws_isof_dns_suffix = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("csp.hci.ic.gov");
 
-int aws_credentials_provider_construct_regional_endpoint(
+AWS_STATIC_STRING_FROM_LITERAL(s_endpoint_url_env, "AWS_ENDPOINT_URL");
+AWS_STATIC_STRING_FROM_LITERAL(s_endpoint_url_property, "endpoint_url");
+AWS_STATIC_STRING_FROM_LITERAL(s_services_property, "services");
+
+struct aws_string *s_get_override_endpoint(
+    struct aws_allocator *allocator,
+    const struct aws_string *override_service_name_env,
+    const struct aws_string *override_service_name_property,
+    const struct aws_profile_collection *profile_collection,
+    const struct aws_profile *profile) {
+
+    struct aws_byte_cursor url_env_cursor = aws_byte_cursor_from_string(s_endpoint_url_env);
+    struct aws_byte_cursor underscore_cursor = aws_byte_cursor_from_c_str("_");
+    struct aws_byte_cursor service_name_env_cursor = aws_byte_cursor_from_string(override_service_name_env);
+
+    struct aws_string *out_endpoint = NULL;
+    struct aws_string *service_endpoint_str = NULL;
+
+    struct aws_byte_buf service_endpoint_buf;
+    AWS_ZERO_STRUCT(service_endpoint_buf);
+    /* Check AWS_ENDPOINT_URL_SERVICENAME variable */
+    aws_byte_buf_init(&service_endpoint_buf, allocator, 10);
+    if (aws_byte_buf_append_dynamic(&service_endpoint_buf, &url_env_cursor) ||
+        aws_byte_buf_append_dynamic(&service_endpoint_buf, &underscore_cursor) ||
+        aws_byte_buf_append_dynamic(&service_endpoint_buf, &service_name_env_cursor)) {
+        goto on_finish;
+    }
+    service_endpoint_str = aws_string_new_from_buf(allocator, &service_endpoint_buf);
+
+    if (aws_get_environment_value(allocator, service_endpoint_str, &out_endpoint) == AWS_OP_SUCCESS &&
+        out_endpoint != NULL) {
+        goto on_finish;
+    }
+
+    /* Check AWS_ENDPOINT_URL variable */
+    if (aws_get_environment_value(allocator, s_endpoint_url_env, &out_endpoint) == AWS_OP_SUCCESS &&
```

Comment:
copy/paste bug? setting `out_endpoint` twice, which will leak the value from the first call

Reply:
oh nevermind if != NULL goto finish